### PR TITLE
[tests] mining functional tests (including regression test for submitblock)

### PR DESCRIPTION
**Pull request by jnewbery**
This PR refactors the getblocktemplate_proposals.py functional test into a more general mining test, which covers the `submitblock` RPC as well as the `getblocktemplate` propose mode.

Motivation was to add a regression test for #10146

This test removes all the manual block/transaction construction from the test case and uses the mininode classes to achieve the same tests. Result is less lines of code, and hopefully clearer and easier for people to extend/maintain since we're no longer hacking bits in a serialised block.

@gmaxwell since he wrote the fix for #10146 

EDIT: there are lots of commits in this PR to aid reviewers. They can be squashed down into larger commits before merging.

### CODE REVIEW

File: test/functional/getblocktemplate_proposals.py [hunk 1]
```diff
@@ -82,13 +83,13 @@ def run_test(self):
...
-            hexoutval = b2x(pack('<Q', tmpl['coinbasevalue']))
-            tmpl['coinbasetxn'] = {'data': '01000000' + '01' + '0000000000000000000000000000000000000000000000000000000000000000ffffffff' + ('%02x' % (len(rawcoinbase),)) + hexcoinbase + 'fffffffe' + '01' + hexoutval + '00' + '00000000'}
-        txlist = list(bytearray(a2b_hex(a['data'])) for a in (tmpl['coinbasetxn'],) + tuple(tmpl['transactions']))
+        assert 'coinbasetxn' not in tmpl
+
+        coinbase_tx = create_coinbase(height=int(tmpl["height"]) + 1)
+        # sequence numbers must not be max for nLockTime to have effect
+        coinbase_tx.vin[0].nSequence = 2 ** 32 - 2
+        tmpl['coinbasetxn'] = {'data': coinbase_tx.serialize()}
+        txlist = [bytearray(coinbase_tx.serialize())]
```
Comment (jimmysong) [inline test/functional/getblocktemplate_proposals.py]: is `tmpl['transactions']` expected to be empty?

File: test/functional/mining.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+def b2x(b):
+    return b2a_hex(b).decode('ascii')
+
+def assert_template(node, block, expect, rehash=True):
+    if rehash:
+        block.hashMerkleRoot = block.calc_merkle_root()
+    rsp = node.getblocktemplate({'data': b2x(block.serialize()), 'mode': 'proposal'})
+    assert_equal(rsp, expect)
+
+class GetBlockTemplateProposalTest(BitcoinTestFramework):
```
Comment (jimmysong) [inline test/functional/mining.py]: rename the test to `MiningTest`?

File: test/functional/mining.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        bad_block.nTime = 0
+        assert_template(node, bad_block, 'time-too-old')
+
+        self.log.info("getblocktemplate: Test not best block")
+        bad_block = copy.deepcopy(block)
+        bad_block.hashPrevBlock = 123
+        assert_template(node, bad_block, 'inconclusive-not-best-prevblk')
+
+if __name__ == '__main__':
+    GetBlockTemplateProposalTest().main()
```
Comment (jimmysong) [inline test/functional/mining.py]: `MiningTest`?

**Review by jimmysong — COMMENTED** [review]
utACK df93f6cc8d80e78eb8758a7d6b5148df9bc3afa9

Gets rid of a lot of duplicate logic. A few nits.

File: test/functional/mining.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+# Copyright (c) 2014-2016 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test block proposals with getblocktemplate."""
```
Comment (jnewbery) [inline test/functional/mining.py]: Yes, thanks.

File: test/functional/mining.py (hunk 2 above)
Comment (jnewbery) [inline test/functional/mining.py]: agreed

File: test/functional/mining.py (hunk 3 above)
Comment (jnewbery) [inline test/functional/mining.py]: yep

File: test/functional/getblocktemplate_proposals.py (hunk 1 above)
Comment (jnewbery) [inline test/functional/getblocktemplate_proposals.py]: yes, I believe so

**Review by jimmysong — APPROVED** [review]
ACK 8d80db230b3a07680cec1c227ff84dbe0678cf45

**Comment by jnewbery** [comment]
@gmaxwell can I convince you to take a look at this? It adds a regression test for your fix in #10146

**Comment by gmaxwell** [comment]
Concept ACK. I am testing the test.

**Comment by jnewbery** [comment]
No great rush for this but it'd be good to get it in soonish. It includes a regression test for #10146, which was backported to 0.14.1. Other benefits:

- tidies up test code
- increases coverage of mining RPCs

**Comment by jnewbery** [comment]
rebased

**Comment by laanwj** [comment]
utACK 11ba8e9

**Pull request merged**